homeassistant/components/tplink/light.py

```diff
@@ -355,47 +322,39 @@ def get_light_state(self, light_features: LightFeatures) -> LightState:
 
         # Tell the device to set the states.
         self._is_setting_light_state = True
-        try:
-            await self.hass.async_add_executor_job(
-                self.set_light_state, old_light_state, new_light_state
-            )
-            self._is_available = True
-            self._is_setting_light_state = False
-            return
-        except (SmartDeviceException, OSError):
-            pass
 
-        try:
-            _LOGGER.debug("Retrying setting light state")
-            await self.hass.async_add_executor_job(
-                self.set_light_state, old_light_state, new_light_state
-            )
-            self._is_available = True
-        except (SmartDeviceException, OSError) as ex:
-            self._is_available = False
-            _LOGGER.warning("Could not set data for %s: %s", self.smartbulb.host, ex)
+        # TODO: this used to be two separate calls to set_light_state and it was simply confusing..
+        for t in range(5):
+            try:
+                await self.set_light_state(old_light_state, new_light_state)
+                self._is_available = True
+                self._is_setting_light_state = False
+                return
+            except (SmartDeviceException, OSError) as ex:
+                _LOGGER.debug("Got error while setting the state, retrying: %s", ex)
+
 
+        _LOGGER.warning("Could not set state for %s", self.smartbulb.host)
         self._is_setting_light_state = False
```

Comment:
You can put this in a `finally` as part of the `try…except`

Reply:
Finally would be executed also in the case where setting the state succeeded after a retry.

Reply:
Right, which it also sets it to False ?

Reply:
There is a return inside the try-block, so in case of success that warning will never be hit.

Reply:
Just move the setting of `_is_setting_light_state` to `False` to finally, not the warning.

Reply:
That will work after removing the retry logic, otherwise it would still require some extra logic to avoid setting it to false if there are still retries left.